internal/util/functions/functions_test.go

```diff
@@ -0,0 +1,198 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package functions_test
+
+import (
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+	"testing"
+
+	"github.com/GoogleContainerTools/kpt/internal/util/functions"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestReconcileFunctions(t *testing.T) {
+	d, err := ioutil.TempDir("", "kpt")
+	if !assert.NoError(t, err) {
+		t.FailNow()
+	}
+
+	// write the Kptfile
+	kf := fmt.Sprintf(`apiVersion: kpt.dev/v1alpha1
+kind: Kptfile
+metadata:
+  name: my-pkg
+functions:
+  starlarkFunctions:
+  - name: func
+    path: %s
+`, filepath.Join(d, "reconcile.star"))
+	err = ioutil.WriteFile(filepath.Join(d, "Kptfile"),
+		[]byte(kf), 0600)
+	if !assert.NoError(t, err) {
+		t.FailNow()
+	}
+
+	deploy := `apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: nginx-deployment
+spec:
+  template:
+    spec:
+      containers:
+      - name: nginx
+        # head comment
+        image: nginx:1.8.1 # {"$ref": "#/definitions/io.k8s.cli.substitutions.image"}
+`
+	err = ioutil.WriteFile(filepath.Join(d, "deploy.yaml"),
+		[]byte(deploy), 0600)
+	if !assert.NoError(t, err) {
+		t.FailNow()
+	}
+
+	fn := `
+# set the foo annotation on each resource
+def run(r):
+  for resource in r:
+    resource["metadata"]["annotations"]["foo"] = "bar"
+
+run(resourceList["items"])
+`
+	err = ioutil.WriteFile(filepath.Join(d, "reconcile.star"),
+		[]byte(fn), 0600)
+	if !assert.NoError(t, err) {
+		t.FailNow()
+	}
+
+	err = functions.ReconcileFunctions(d)
+	if !assert.NoError(t, err) {
+		t.FailNow()
+	}
+
+	actual, err := ioutil.ReadFile(filepath.Join(d, "deploy.yaml"))
+	if !assert.NoError(t, err) {
+		t.FailNow()
+	}
+
+	expected := `apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: nginx-deployment
+  annotations:
+    foo: bar
+spec:
+  template:
+    spec:
+      containers:
+      - name: nginx
+        # head comment
+        image: nginx:1.8.1 # {"$ref": "#/definitions/io.k8s.cli.substitutions.image"}
+`
+
+	if !assert.Equal(t, expected, string(actual)) {
+		t.FailNow()
+	}
+}
+
+func TestReconcileFunctionsImplicit(t *testing.T) {
```

Comment:
Might be trivial, but shouldn't we add a case for autoRunContainer: true

Reply:
Testing running functions isn't easy due to the DiD problem.  I could delete the option from this PR.